tests/test.rs

```diff
@@ -454,3 +454,18 @@ fn triple_nested_structure() {
 
     assert_eq!(expected_values, nested_objects);
 }
+
+#[test]
+fn uncopyable_iterator() {
+    struct Foo {};
+    impl Iterator for Foo {
+        type Item = ();
+        fn next(&mut self) -> Option<()> {
+            None
+        }
+    }
+    let _ = comp! {
+        x;
+        for x in Foo {};
+    };
+}
```

Comment:
Here's a test for checking unsized iterators, based on your test:

```rust
#[test]
fn unsized_iterator() {
    struct Foo {};

    impl Iterator for Foo {
        type Item = ();

        fn next(&mut self) -> Option<()> {
            None
        }
    }

    fn iterate_as_unsized<T>(iter: &T)
    where
        T: Iterator + ?Sized,
    {
        let _ = comp!(x; for x in *iter);
    }

    iterate_as_unsized(&Foo {})
}
```

except that trying this, it turns out that we get a really interesting error:

```
error[E0277]: the size for values of type `T` cannot be known at compilation time
   --> tests/test.rs:489:17
    |
489 |         let _ = comp!(x; for x in *iter);
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `std::marker::Sized` is not implemented for `T`
    = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
    = help: consider adding a `where T: std::marker::Sized` bound
    = note: all local variables must have a statically known size
    = help: unsized locals are gated as an unstable feature
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0277]: the size for values of type `T` cannot be known at compilation time
   --> tests/test.rs:489:17
    |
489 |         let _ = comp!(x; for x in *iter);
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `std::marker::Sized` is not implemented for `T`
    = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
    = help: consider adding a `where T: std::marker::Sized` bound
    = note: required because of the requirements on the impl of `std::iter::IntoIterator` for `T`
    = note: required by `py_comp::assert_impl_into_iter`
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

Which says two really interesting things:
1. "unsized locals are gated as an unstable feature"
   because the closures in the macro capture the iterator by value (with the `move`) and then assign it to the local `iter`, we surprisingly hit a limit in the current type system and memory model. So... at least this way, it may be impossible to accept unsized iterators. (can closures be unsized?)
2. "the trait `std::marker::Sized` is not implemented for `T`" ... "required because of the requirements on the impl of `std::iter::IntoIterator` for `T`"
   Looks like It's a builtin feature that iterators can't be unsized at all. Am i reading this correctly?

maybe we should drop the `?Sized` bound from `assert_impl_into_iter` altogether?

Reply:
IntoIterator has a blanket impl for Iterator, but it's not implemented for `Iterator + ?Sized` because it wouldn't be able to access `Iterator::Item` for trait objects.

Oh, right. IntoIterator also has an associated type, so it's not object-safe either...